nexus/src/app/sagas/instance_create.rs

```diff
@@ -758,64 +735,50 @@ async fn ensure_instance_disk_attach_state(
+    let params = sagactx.saga_params::<DiskAttachParams>()?;
+    let datastore = osagactx.datastore();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+    let project_id = params.project_id;
 
-    let saga_disks = &saga_params.create_params.disks;
-    let instance_name =
-        db::model::Name(saga_params.create_params.identity.name);
+    let disk_name = match params.attach_params {
+        InstanceDiskAttachment::Create(create_params) => {
+            db::model::Name(create_params.identity.name)
+        }
+        InstanceDiskAttachment::Attach(attach_params) => {
+            db::model::Name(attach_params.name)
+        }
+    };
 
-    if disk_index >= saga_disks.len() {
-        return Ok(());
-    }
+    let (.., authz_instance, _db_instance) =
+        LookupPath::new(&opctx, &datastore)
+            .instance_id(instance_id)
+            .fetch()
+            .await
+            .map_err(ActionError::action_failed)?;
 
-    let disk = &saga_disks[disk_index];
+    let (.., authz_disk, _db_disk) = LookupPath::new(&opctx, &datastore)
+        .project_id(project_id)
+        .disk_name(&disk_name)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
 
-    // TODO-correctness TODO-security It's not correct to re-resolve the
-    // organization and project names now.  See oxidecomputer/omicron#1536.
-    let organization_name: db::model::Name =
-        saga_params.organization_name.clone().into();
-    let project_name: db::model::Name = saga_params.project_name.clone().into();
-
-    match disk {
-        params::InstanceDiskAttachment::Create(_) => {
-            // TODO grab disks created in sic_create_disks_for_instance
-            return Err(ActionError::action_failed(Error::invalid_request(
-                "creating disks while creating an instance not supported",
-            )));
-        }
-        params::InstanceDiskAttachment::Attach(instance_disk_attach) => {
-            let disk_name: db::model::Name =
-                instance_disk_attach.name.clone().into();
-
-            if attached {
-                osagactx
-                    .nexus()
-                    .instance_attach_disk(
-                        &opctx,
-                        &organization_name,
-                        &project_name,
-                        &instance_name,
-                        &disk_name,
-                    )
-                    .await
-            } else {
-                osagactx
-                    .nexus()
-                    .instance_detach_disk(
-                        &opctx,
-                        &organization_name,
-                        &project_name,
-                        &instance_name,
-                        &disk_name,
-                    )
-                    .await
-            }
+    if attached {
+        datastore
+            .instance_attach_disk(
```

Comment:
Just for my understanding, as I'm not super familiar with this code: Why change the call from using `osagactx.nexus().instance_XXX` to `datastore.instance_XXX`?

It seems you basically duplicated the methods from nexus/src/app/instance.rs

Reply:
Yeah, this was a misunderstanding on my part. I'd mistakenly thought that functions off of `osagactx.nexus()` spawned sagas when that's clearly just an implementation detail. I'll change it back to what it was.

Reply:
Actually a correction to that: `nexus().instance_attach_disk` isn't the best form. It takes names and does lookups which are resolved before calling the datastore method. You're right that we should reuse `instance_attach_disk` in `/nexus/src/app/instance` but ideally we'd have a different form that takes pre-resolved arguments like the datastore method. 

Ideally we want to resolve as many names as possible at the very start of a saga (or better, before the saga is invoked). 

Do folks think I should change it back or is this form acceptable until we're able to follow up with a better interface for attaching? cc @smklein, @davepacheco. 

This sort of plays into [my previous comment](https://github.com/oxidecomputer/omicron/pull/1873/files/3927efedfa8e89c41460ef33697bdbdefb34ad4d#r1007193729).

Reply:
At this point I'd personally err on the side of "doing less name lookups" - the boundary between nexus and datastore is really thin from the point-of-view of sagas. The datastore itself should provide necessary encapsulation for safe DB access, but forcing sagas to go through nexus isn't adding much IMO.

Reply:
A few things:

- Yes, I think at some point we discussed having the Nexus-level functions accept `LookupPath` objects rather than "names" so that we could use them from places that had ids or names.
- Yes, I think sagas should use the datastore freely when that's the right level of abstraction.
- At the same time, it might also be appropriate to use the Nexus-level functions when there's real application-level logic that has to happen.  For example, I think it'd be better for sagas that need to update instance state to use the app-level instance_set_runtime().  In this specific case, it seems like the app-level wrapper doesn't do anything _but_ the name resolution and then call the datastore function.  So it seems fine.